mlir/lib/Dialect/Rock/Transforms/GemmOutputSwizzle.cpp

```diff
@@ -0,0 +1,571 @@
+    auto tidIterMergeAttr = tidIterMerge.get();
+
+    auto tidIterFlatten =
+        TopDownTMBuilder::below(tidIterMerge, tidIterMergeAttr);
+    tidIterFlatten.unmerge("flattenBlock", 0, {"iter", "tid", "numElements"},
+                           {iter, blockSize, elementsWrittenPerThread});
+    auto tidIterFlattenAttr = tidIterFlatten.get();
+
+    SmallVector<Attribute> transformAttrs;
+    transformAttrs.push_back(tidIterMergeAttr);
+    transformAttrs.push_back(tidIterFlattenAttr);
+
+    ArrayAttr ldsRead = b.getArrayAttr(transformAttrs);
+    auto ldsBufferForLoad = transform(b, typedBuffer, ldsRead);
+
+    // LDS barrier.
+    b.create<LDSBarrierOp>(loc);
+
+    b.create<ThreadwiseReadIntoOp>(loc, ldsBufferForLoad, finalC,
+                                   b.getArrayAttr({}), ValueRange{tid},
+                                   forceUnroll, useIndexDiffs);
+
+    SmallVector<int64_t, 5> bidGridLengths;
+    SmallVector<StringRef, 5> bidGridOrder;
+    int64_t mBlocks = M / mPerBlock;
+    int64_t nBlocks = N / nPerBlock;
+    if (M % mPerBlock != 0 || N % nPerBlock != 0) {
+      return failure();
+    }
+    if ((mBlocks * nBlocks) != gridSize) {
+      return failure();
+    }
+    bool isAttention = op.getExtraIndices().size() == 3;
+    LLVM_DEBUG(llvm::dbgs() << "isAttention: " << isAttention << "\n");
+    if (isAttention) {
+      // attention
+      dataPerThread = dataPerThread * mBlocks;
+      mBlocks = 1;
+      bidGridLengths = {G, nBlocks, blockSize, dataPerThread};
+      bidGridOrder = {"g_block", "n_block", "tid", "iter"};
+    } else {
+      bidGridLengths = {G, mBlocks, nBlocks, blockSize, dataPerThread};
+      bidGridOrder = {"g_block", "m_block", "n_block", "tid", "iter"};
+    }
+
+    // Save to memory
+    // Create views as gridwise sub-tile of C
+    TopDownTMBuilder createMblock(b, bidGridOrder, bidGridLengths, loc);
+    if (isAttention) {
+      createMblock.passThrough({"g_block", "n_block", "tid"}, {0, 2, 3},
+                               {"g_block", "n_block", "tid"});
+      createMblock.merge({"m_block", "iter"}, {1, 4}, "iter",
+                         {mBlocks, dataPerThread / mBlocks});
+    } else {
+      createMblock.passThrough(bidGridOrder, {0, 1, 2, 3, 4}, bidGridOrder);
+    }
+    auto createMblockAttr = createMblock.get();
+
+    TopDownTMBuilder tidIterMergeMem =
+        TopDownTMBuilder::below(createMblock, createMblockAttr);
+    tidIterMergeMem.passThrough({"g_block", "m_block", "n_block", "tid"});
+    tidIterMergeMem.merge({"iter", "numElements"}, {4, 5}, "iter",
+                          {iter, elementsWrittenPerThread});
+    auto tidIterMergeMemAttr = tidIterMergeMem.get();
+
+    TopDownTMBuilder tidIterFlattenMem =
+        TopDownTMBuilder::below(tidIterMergeMem, tidIterMergeMemAttr);
+    tidIterFlattenMem.passThrough({"g_block", "m_block", "n_block"});
+    tidIterFlattenMem.unmerge("flattenBlock", 3, {"iter", "tid", "numElements"},
```

Comment:
Will it be possible factor out that does the transformation of tid, iter --> iterVec, tid, VecItem + flattenting?
I think logic would be to passthrough everything else.